### PR TITLE
Migrate 8 very similar FFI `run-make` tests to rmake

### PR DESCRIPTION
**Pull request by Oneirical**
Part of #121876 and the associated [Google Summer of Code project](https://blog.rust-lang.org/2024/05/01/gsoc-2024-selected-projects.html).

There are some more of these, but while the code is almost always the same, I want to keep the number reasonable so my doc comments can be inspected for potential inaccuracies. Tell me if 8 is too much, I can cut this down.

For the tracking issue:

- issue-25581
- extern-fn-with-extern-types
- extern-fn-struct-passing-abi
- longjmp-across-rust
- static-extern-type
- extern-fn-explicit-align
- extern-fn-with-packed-struct
- extern-fn-mangle

### CODE REVIEW

**Comment by Oneirical** [comment]
`error adding symbols: archive has no index; run ranlib to add one`

Yikes. According to some quick research, this has something to do with the AR command... I've seen a suggestion to use `llvm-ar` [here](https://github.com/emscripten-core/emscripten/issues/9329). Maybe this could be pertinent the same way we're replacing `nm` with `llvm-readobj`.

**Comment by jieyouxu** [comment]
`build_native_static_lib` IIRC was present in another PR. Should we mark this PR as blocked on the other PR being merged?
@rustbot author

**Comment by jieyouxu** [comment]
> `error adding symbols: archive has no index; run ranlib to add one`
> 
> Yikes. According to some quick research, this has something to do with the AR command... I've seen a suggestion to use `llvm-ar` [here](https://github.com/emscripten-core/emscripten/issues/9329). Maybe this could be pertinent the same way we're replacing `nm` with `llvm-readobj`.

I don't mind switching back from `ar` crate to `llvm-ar`, seeing as we would reasonably expect `llvm-ar` to be present in the environment for consistency across multiple environments. I think it's worth a try.

**Comment by Oneirical** [comment]
Very splendid, `llvm-ar` looks like it works. Maybe there are some try-jobs you may be interested in running.
@rustbot review

File: src/tools/run-make-support/src/lib.rs [hunk 1]
```diff
@@ -317,6 +317,26 @@ pub fn count_regex_matches_in_files_with_extension(re: &regex::Regex, ext: &str)
...
+        cc().arg("-c").out_exe(&obj_file).input(src).run();
+    } else {
+        cc().arg("-v").arg("-c").out_exe(&obj_file).input(src).run();
+    };
+    let mut obj_file = PathBuf::from(format!("{lib_name}.o"));
+    if is_msvc() {
+        obj_file.set_extension("");
+        obj_file.set_extension("obj");
+    }
+    llvm_ar().obj_to_ar().output_input(&lib_path, &obj_file).run();
```
Comment (jieyouxu) [inline src/tools/run-make-support/src/lib.rs:336]: Question: similar concern as https://github.com/rust-lang/rust/pull/126208#discussion_r1677102091, maybe let's block this PR on that?

**Pull request merged**